### PR TITLE
Lots of changes. Lots of anti-patterns (probably lol). Just got throu…

### PR DESCRIPTION
**Pull request by wsievern**
…gh a Git tutorial, I hope I'm doing this correctly! Will push smaller changes from now on.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -7,10 +8,14 @@
 
 Window.size = (500, 1000)
 
+LabelBase.register(name='GillSans',
```
Comment (emptyopen) [inline main.py:11]: nit: generally you want to be consistent with single quotes or double quotes for strings throughout a project. you can easily solve this with an automatic linter but no big deal.

File: models/piece.py [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
         self.filename = filename
         self.displayname = displayname
         self.bpm = bpm
+        self.info = info
 
 piece_collection = (
-        Piece("data/plies/Chopin_ Waltz No. 7 in C-Sharp Minor, Op. 64 No. 2.mp3", "Waltz in C-Sharp Minor", 70),
-        Piece("data/plies/Frederic Chopin - Mazurka Op. 33, No. 4.mp3", "Mazurka Op. 33, No. 4", 80),
-        Piece("data/plies/Frederic Chopin - Waltz in A Minor, B.150.mp3", "Waltz in A Minor", 90),
+        Piece("/Users/willsievern/Documents/GitHub/BalletBot/data/plies/Chopin_ Waltz No. 7 in C-Sharp Minor, "
```
Comment (emptyopen) [inline models/piece.py:9]: sometimes when you have an idea to do something but it would take too long to do it at the moment, you can track your thoughts as TODOs. So here above these pieces you could put a comment like:
```
# TODO: automatically retrieve file names from directory with os.glob
```

File: models/sampler.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+class Sampler(object):
+    """ Sampler used to play, stop and mix multiple sounds.
+        .. warning:: A single sampler instance should be used at a time.
+    """
+
+    def __init__(self, sr=22050, backend='sounddevice', timeout=1):
+        """
+        :param int sr: samplerate used - all sounds added to the sampler will automatically be resampled if needed (- his can be a CPU consumming task, try to use sound with all identical sampling rate if possible.
+        :param str backend: backend used for playing sound. Can be either 'sounddevice' or 'dummy'.
```
Comment (emptyopen) [inline models/sampler.py:17]: probably best to make this an Enum

File: models/sampler.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        self.sounds = []
+
+        self.chunks = Queue(1)
+        self.chunk_available = Condition()
+        self.is_done = Event()  # new event to prevent play to be called again before the sound is actually played
+        self.timeout = timeout  # timeout value for graceful exit of the BackendStream
+
+        if backend == 'dummy':
+            from .dummy_stream import DummyStream
+            self.BackendStream = DummyStream
```
Comment (emptyopen) [inline models/sampler.py:29]: variables should be underscore_case

File: models/sampler.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        self.sr = sr
+        self.sounds = []
+
+        self.chunks = Queue(1)
+        self.chunk_available = Condition()
+        self.is_done = Event()  # new event to prevent play to be called again before the sound is actually played
+        self.timeout = timeout  # timeout value for graceful exit of the BackendStream
+
+        if backend == 'dummy':
+            from .dummy_stream import DummyStream
```
Comment (emptyopen) [inline models/sampler.py:28]: I think it's bad practice to conditionally import, should have all imports at the top of file

File: models/sampler.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        """
+        :param int sr: samplerate used - all sounds added to the sampler will automatically be resampled if needed (- his can be a CPU consumming task, try to use sound with all identical sampling rate if possible.
+        :param str backend: backend used for playing sound. Can be either 'sounddevice' or 'dummy'.
+        """
+        self.sr = sr
+        self.sounds = []
+
+        self.chunks = Queue(1)
+        self.chunk_available = Condition()
+        self.is_done = Event()  # new event to prevent play to be called again before the sound is actually played
```
Comment (emptyopen) [inline models/sampler.py:24]: seems like you might not be using threading as expected here, perhaps a topic we can explore 1:1

File: models/sampler.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+        t = Thread(target=chunks_producer)
+        t.start()
+
+        with self.BackendStream(samplerate=self.sr, channels=1) as stream:
+            while self.running:
+                try:
+                    stream.write(self.chunks.get(timeout=self.timeout))  # timeout so stream.write() thread can exit
+                except Empty:
+                    self.running = False  # let play_thread exit
```
Comment (emptyopen) [inline models/sampler.py:116]: might want to move this out of the "with" block, to turn running off in the event of some other exception

File: models/sampler.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
+from queue import Empty, Queue
```
Comment (emptyopen) [inline models/sampler.py:1]: generally "models" are more of a data representation. This is more of a service kind of layer, might want to rename it to like "SamplerService" and put it in a new "service" directory, alongside "models"

File: models/sound.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+import numpy
+import librosa
+
+
+class Sound(object):
+    """ Sound object which can be read chunk by chunk. """
+
+    def __init__(self, y, sr, chunk_size=1024):
+        """
+        :param array y: numpy array representing the audio data
```
Comment (emptyopen) [inline models/sound.py:12]: probably better variable than y

File: models/sound.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+        :param int sr: samplerate
+        :param int chunk_size: size of each chunk (default 1024)
+        You also have access two properties:
```
Comment (emptyopen) [inline models/sound.py:15]: "to three additional"

File: models/sound.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+            * pitch_shift (real time pitch shifting)
+            * time_stretch (real time time-scale without pitch modification)
+            * volume (real time volume level)
+        Both can be tweaked simultaneously.
```
Comment (emptyopen) [inline models/sound.py:19]: Both -> these

File: models/sound.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        """
+        self.y, self.sr = y.astype(dtype='float32'), sr
```
Comment (emptyopen) [inline models/sound.py:21]: better to separate lines

File: models/sound.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        self._init_stretching()
+
+        self.loop = False
+
+        # Effect properties
+        self.pitch_shift = 0
+        self.stretch_factor = 1.0
+        self.volume = 1.0
+
+    def resample(self, target_sr):
```
Comment (emptyopen) [inline models/sound.py:33]: can make this a static method

File: models/sound.py [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+        if value and hasattr(self, '_it'):
+            self._reset()
+
+        self._playing = value
+
+    @property
+    def chunks(self):
+        """ Returns a chunk iterator over the sound. """
+        if not hasattr(self, '_it'):
+            class ChunkIterator(object):
```
Comment (emptyopen) [inline models/sound.py:70]: might be cleaner to define this in its own file

File: models/sound.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+                next = __next__
+
+            self._it = ChunkIterator()
+
+        return self._it
+
+    def _next_chunk(self):
+        chunk = self._time_stretcher(self.stretch_factor)
+
+        if numpy.round(self.pitch_shift, 1) != 0:
```
Comment (emptyopen) [inline models/sound.py:94]: I think you can just remove "!= 0" as non zero is truthy

File: models/sound.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+        chunk *= self.volume
+
+        if len(chunk) != self.chunk_size:
+            raise StopIteration
+
+        return chunk
+
+    # Effects
+
+    def pitch_shifter(self, chunk, shift):
```
Comment (emptyopen) [inline models/sound.py:105]: nit: might be more consistent to name function with active tense like "shift_pitch"

File: models/sound.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+        """ Pitch-Shift the given chunk by shift semi-tones. """
+        freq = numpy.fft.rfft(chunk)
+
+        N = len(freq)
```
Comment (emptyopen) [inline models/sound.py:109]: in general longer variable names can save a lot of readability headaches for yourself in the future, even if it's obvious to you now.

File: models/sound.py [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+    @property
+    def stretch_factor(self):
+        return self._sf
+
+    @stretch_factor.setter
+    def stretch_factor(self, value):
+        self._sf = value
+        self._zero_padding()
+
+    def _time_stretcher(self, stretch_factor):
```
Comment (emptyopen) [inline models/sound.py:155]: for all of these math functions especially, it can be good to have a separate test file (in general testing is a critical skill in SW engineering) to test corner cases and such. even if you end up swapping out libraries, it's a quick way to make sure you don't break anything

File: screens/tap_tempo/tap_tempo.py [hunk 19]
```diff
@@ -85,31 +55,33 @@ def tap_estimate(self):
...
-time = np.arange(0,len(y_slow))/sr
-fig, ax = plt.subplots()
-ax.plot(time,y_slow)
-ax.set(xlabel='Time(s)',ylabel='sound amplitude')
-plt.show()'''
+        if len(self.time_list) > 1:
+            self.display_tempo.text = (str(TapTempo.tempo))
+        if len(self.time_list) > 2 and TapTempo.tempo < self.app.root.ids.playback.ids.bpm_slider.max:
+            self.app.root.ids.playback.ids.bpm_slider.value = TapTempo.tempo
+            self.app.root.ids.playback.ids.display_slider_value.text = str(round(TapTempo.tempo * (mem.selected_piece_bpm ** -1), 1)) + "x : " + (str(round(TapTempo.tempo, 0)))
```
Comment (emptyopen) [inline screens/tap_tempo/tap_tempo.py:71]: I'd recommend using f-strings:

`str(round(TapTempo.tempo * (mem.selected_piece_bpm ** -1), 1)) + "x : " + (str(round(TapTempo.tempo, 0)))`
vs
`f"{round(TapTempo.tempo * (mem.selected_piece_bpm ** -1), 1)}x : {round(TapTempo.tempo, 0)}"`
or even
variable1 = round(TapTempo.tempo * (mem.selected_piece_bpm ** -1), 1)
variable2 = round(TapTempo.tempo, 0)
`f"{variable1}x : {variable2}"`

**Review by emptyopen — COMMENTED** [review]
Added some comments!
